[WIP] [v1.x] Create tool for building source archives

josephevans opened the pull request:
## Description ##
This PR introduces a script that automates creating a release source archive and handles some licensing issues by removing some files that were identified to be non-compliant, and introducing a Apache RAT license check in the source generation process to ensure all Apache MXNet code is compliant.

Zha0q1 commented:
@josephevans Does this same tool also work for master?

sandeep-krishnamurthy commented on tools/create_source_archive.sh:
```diff
@@ -0,0 +1,82 @@
+echo "Removing unwanted artifacts..."
+#### IMPORTANT ####
+# Remove artifacts which do not comply with the Apache Licensing Policy
+rm -rf R-package
```
Can you make this part of another file say another text file. so it is easy to audit for someone during voting without looking at a code file.

josephevans replied:
Sure, any recommendations on where this file should live in the repo?

josephevans replied:
Put them in `tools/source-exclude-artifacts.txt` for now, please let me know if there is a better location.

josephevans commented:
> @josephevans Does this same tool also work for master?

There will be some slight modifications for master (there's no R-package in master, which we remove in v1.x.)

josephevans commented:
I was going to remove `mx-theme` references in the LICENSE file, but it looks like @szha already has them removed in #19873. Should we also remove references to `mathjax` (from mkldnn docs) in LICENSE since we will no longer be shipping it out in the source archive? @leezu

leezu commented on tools/create_source_archive.sh:
```diff
@@ -0,0 +1,86 @@
+# Remove other artifacts we do not want contained in the source archive
+rm -rf .DS_Store
+rm -rf CODEOWNERS
+find . -name ".git*" -print0 | xargs -0 rm -rf
```
How about passing `--exclude-vcs` to `tar` instead?

josephevans replied:
Sure, but it won't remove `.github` (based on https://www.gnu.org/software/tar/manual/html_node/exclude.html) - I can also add that to the list on the line above.

leezu commented on tools/source-exclude-artifacts.txt:
```diff
@@ -0,0 +1,24 @@
+#  not be included in source release archives due to licensing
+#  restrictions.
+
+R-package
```
What is the concern about the R frontend?

josephevans replied:
It was already excluded in the release runbook (see https://cwiki.apache.org/confluence/display/MXNET/Release+Process#ReleaseProcess-Step1.10.Createartifactsforthereleaseandpushtothedistfolder). Has this licensing issue been resolved? Can it be removed from the exclude list?

leezu replied:
I don't know why it was listed there. That is very dubious, because in that case R-package has never been an official feature of MXNet and we may just delete the code directly as it's no longer maintained anyways..

josephevans replied:
Verified with @szha, we still need to remove R-package, so this should be correct.

leezu commented:
> Should we also remove references to mathjax (from mkldnn docs) in LICENSE since we will no longer be shipping it out in the source archive? 

One idea is to automatically remove lines fro LICENSE that match a line in `tools/source-exclude-artifacts.txt`?

josephevans commented on tools/create_source_archive.sh:
```diff
@@ -0,0 +1,86 @@
+
+# clone the repo and checkout the tag
+echo "Cloning the MXNet repository..."
+git clone -b $MXNET_BRANCH https://github.com/apache/incubator-mxnet.git $SRCDIR
```
I was assuming this script was going to be run manually when a new release is cut, do we really care about shedding a few seconds off here?

josephevans replied:
Updated.

josephevans commented:
> > Should we also remove references to mathjax (from mkldnn docs) in LICENSE since we will no longer be shipping it out in the source archive?
> 
> One idea is to automatically remove lines fro LICENSE that match a line in `tools/source-exclude-artifacts.txt`?

Since it will no longer be distributed in the source archive, and it shouldn't be distributed in binary releases, why does it need to be in at all? I can definitely exclude lines that match our source-exclude-artifacts.txt list, but I'm not sure I understand the reasoning behind requiring it in the first place.

szha commented:
It would be great to add a check for all files listed in license and warn if any doesn't exist

josephevans commented:
> It would be great to add a check for all files listed in license and warn if any doesn't exist

Added this to the script. Found 2 items in the LICENSE file that are no longer in the source tree and removed them.,

josephevans commented:
> > Should we also remove references to mathjax (from mkldnn docs) in LICENSE since we will no longer be shipping it out in the source archive?
> 
> One idea is to automatically remove lines fro LICENSE that match a line in `tools/source-exclude-artifacts.txt`?

OK, I updated to remove lines in LICENSE for stuff we remove the source archive.
